mountpoint-s3/src/upload.rs

```diff
@@ -52,21 +53,23 @@ pub enum UploadWriteError<E: std::error::Error> {
 /// Manages the upload of an object to S3.
 ///
 /// Wraps a PutObject request and enforces sequential writes.
-pub struct UploadRequest<Client: ObjectClient> {
+pub struct UploadRequest<Client: ObjectClient, Handle> {
     bucket: String,
     key: String,
     next_request_offset: u64,
     request: Client::PutObjectRequest,
+    _handle: Handle,
```

Comment:
I'm not sure the handle should be here -- it doesn't really have anything to do with S3 uploads (it's a filesystem concept). It should probably be back in `fs`. If that gets too messy, we should factor out a `WriteFileHandle` type and change the enum to `FileHandleType::Write(WriteFileHandle)` so that we can put the transition logic onto that struct instead.

So something like
```rust
struct WriteFileHandle {
    upload: UploadRequest<Client, WriteHandle>,
    token: WriteHandle,
}

impl WriteFileHandle {
    fn complete(mut self) -> Result<(), libc::c_int> {
        // ...the common logic that lives in both `fsync` and `release` right now...

        // and explicitly complete the write handle
        self.token.finish_writing()?;
    }
}
```